### PR TITLE
Fix #3076 - issues related to in_offset overflow in Col2Im (2D and 3D) and Im2Col (2D only)

### LINKED ISSUE #3076: Memory access fault for Gemm solvers after #2969
@atamazov this issue does not appear if we revert #2969 

[Observation]:
If we run `vgg16` model on 8 GPU concurrently:
```
python3 micro_benchmarking_pytorch.py --network vgg16 --batch-size 1024 --iterations 100 --distributed_dataparallel --device_ids=0 --rank 0 --world-size 8 --dist-backend nccl --dist-url tcp://127.0.0.1:4332 2>&1 | tee log_rank0 &
python3 micro_benchmarking_pytorch.py --network vgg16 --batch-size 1024 --iterations 100 --distributed_dataparallel --device_ids=1 --rank 1 --world-size 8 --dist-backend nccl --dist-url tcp://127.0.0.1:4332 2>&1 | tee log_rank1 &
python3 micro_benchmarking_pytorch.py --network vgg16 --batch-size 1024 --iterations 100 --distributed_dataparallel --device_ids=2 --rank 2 --world-size 8 --dist-backend nccl --dist-url tcp://127.0.0.1:4332 2>&1 | tee log_rank2 &
python3 micro_benchmarking_pytorch.py --network vgg16 --batch-size 1024 --iterations 100 --distributed_dataparallel --device_ids=3 --rank 3 --world-size 8 --dist-backend nccl --dist-url tcp://127.0.0.1:4332 2>&1 | tee log_rank3 &
python3 micro_benchmarking_pytorch.py --network vgg16 --batch-size 1024 --iterations 100 --distributed_dataparallel --device_ids=4 --rank 4 --world-size 8 --dist-backend nccl --dist-url tcp://127.0.0.1:4332 2>&1 | tee log_rank4 &
python3 micro_benchmarking_pytorch.py --network vgg16 --batch-size 1024 --iterations 100 --distributed_dataparallel --device_ids=5 --rank 5 --world-size 8 --dist-backend nccl --dist-url tcp://127.0.0.1:4332 2>&1 | tee log_rank5 &
python3 micro_benchmarking_pytorch.py --network vgg16 --batch-size 1024 --iterations 100 --distributed_dataparallel --device_ids=6 --rank 6 --world-size 8 --dist-backend nccl --dist-url tcp://127.0.0.1:4332 2>&1 | tee log_rank6 &
python3 micro_benchmarking_pytorch.py --network vgg16 --batch-size 1024 --iterations 100 --distributed_dataparallel --device_ids=7 --rank 7 --world-size 8 --dist-backend nccl --dist-url tcp://127.0.0.1:4332 2>&1 | tee log_rank7 &
```
we will oserve
```
MIOpen(HIP): miopenStatus_t miopen::CallGemm(const Handle &, GemmDescriptor, ConstData_t, std::size_t, ConstData_t, std::size_t, Data_t, std::size_t, GemmBackend_t){
MIOpen(HIP): 	"rocBLAS" = rocBLAS
MIOpen(HIP): }
MIOpen(HIP): miopenStatus_t miopen::CallGemm(const Handle &, GemmDescriptor, ConstData_t, std::size_t, ConstData_t, std::size_t, Data_t, std::size_t, GemmBackend_t){
MIOpen(HIP): 	"rocBLAS" = rocBLAS
MIOpen(HIP): }
MIOpen(HIP): miopenStatus_t miopen::CallGemm(const Handle &, GemmDescriptor, ConstData_t, std::size_t, ConstData_t, std::size_t, Data_t, std::size_t, GemmBackend_t){
MIOpen(HIP): 	"rocBLAS" = rocBLAS
MIOpen(HIP): }
MIOpen(HIP): miopenStatus_t miopen::CallGemm(const Handle &, GemmDescriptor, ConstData_t, std::size_t, ConstData_t, std::size_t, Data_t, std::size_t, GemmBackend_t){
MIOpen(HIP): 	"rocBLAS" = rocBLAS
MIOpen(HIP): }s fault by GPU node-2 (Agent handle: 0x8fd5cb0) on address 0x7f90df780000. Reason: Unknown.
E0617 20:33:08.607907 139846246954176 torch/distributed/elastic/multiprocessing/api.py:826] failed (exitcode: -6) local_rank: 4 (pid: 1215) of binary: sb
```

`export MIOPEN_DEBUG_CONV_GEMM=0` can avoid the crash too.

So the running theory:
- Some configs are not found in DB, so MIOpen benchmarks a few solvers, which include the GEMM ones
- The GEMM ones have issue with workspace allocation and cause mem fault (even though they will not be picked anyway)

### PR DESCRIPTION
**Pull request by atamazov**
Intended to resolve #3076. Fixes some issues in `GemmBwdRest`, `GemmFwdRest`, `GemmWrwUniversal`. 

- Switches 2D and 3D Col2Im kernels to _**unsigned**_ index computations math.
  - As a by-product, this resolves point 2 of https://github.com/ROCm/MIOpen/pull/1966/files#r1106928625
- Adds 64-bit index mode to 2D Col2Im
  - [Informative] Note that I do not see any perf drop related to switching to 64-bit index computations, so maybe we can simply drop 32-bit stuff and simplify the code (actually I see very small gain))). However right now let's keep 32-bit stuff to minimize changes (who knows)?
  - Passes in_offset as ulong instead of int to 2D Im2Col
- There is no find-db and KDB related impact because:
  - Modified kernels were renamed, so the updated library won't load existing kernels from persistent cache. Therefore, regeneration of KDBs is not required (but still desired to avoid performance hit, albeit very small).
  - Applicability of the solvers remain the same.

## 🌀 Examples of resolved issues:

Reproducible on 16GB cards.

- `./bin/MIOpenDriver convfp16 -n 16 -c 400 -H 224 -W 224 -k 1 -y 20 -x 20 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -F 2 -i 1 -V 1 -S GemmBwdRest`
  - ch_offset_max = 6855840000 (6.38 G)
  - ORG: Backward Convolution Data FAILED: 0.0681179 > 0.0082
  - FIX: Backward Convolution Data Verifies OK on GPU reference (0.000153848 < 0.0082)
- `./bin/MIOpenDriver convfp16 -n 3000 -c 1 -H 1024 -W 1024 -k 1 -y 3 -x 3 -F 1 -V 1 -i 1 -F 1 -V 1 -S GemmFwdRest`
  - in_offset_max = 3144679424 (2.92871 G)
  - ORG: Forward Convolution FAILED: 0.264809 > 0.0082
  - FIX: Forward Convolution Verifies OK on GPU reference (2.5259e-06 < 0.0082)
- `./bin/MIOpenDriver convfp16 -n 101 -c 1000 -H 160 -W 160 -k 1 -y 3 -x 3 -F 4 -S GemmWrwUniversal -i 1 -V 1`
  - in_offset_max = 2560000000 (2.38419 G)
  - ORG: Memory access fault by GPU node-1 (Agent handle: 0x25e6ef0) on address 0x7f6fb6b93000. Reason: Page not present or supervisor privilege.
  - FIX: Backward Convolution Weights Verifies OK on GPU reference (0.00192734 < 0.0164)

----
[Attribution] @junliume @JehandadKhan
- https://github.com/ROCm/MIOpen/labels/bug
- https://github.com/ROCm/MIOpen/labels/correctness
- https://github.com/ROCm/MIOpen/labels/urgency_blocker
- Proposed reviewers:
  - @JehandadKhan 
  - @junliume
  - @bghimireamd

### CODE REVIEW

File: src/kernels/MIOpenCol2Im2d.cl [hunk 1]
```diff
@@ -33,51 +33,61 @@
...
+                       ? 0
+                       : (im_h - (dilation_h * (wei_h - 1) + 1)) / stride_h + 1;
+    uint end_h   = min(col_h, im_h / stride_h + 1);
+    uint start_w = (im_w < dilation_w * (wei_w - 1) + 1)
+                       ? 0
+                       : (im_w - (dilation_w * (wei_w - 1) + 1)) / stride_w + 1;
+    uint end_w   = min(col_w, im_w / stride_w + 1);
 
-    int ch_offset = im_ch * col_w * col_h * wei_w * wei_h;
+    index_t ch_offset = (index_t)(im_ch * col_w * col_h) * (index_t)(wei_w * wei_h);
```
Comment (atamazov) [inline src/kernels/MIOpenCol2Im2d.cl:79]: This prevents overflow in Col2Im2D. `index_t` depends on value of `MIOPEN_USE_64BIT_INDEX`.

File: src/kernels/MIOpenCol2Im2d.cl [hunk 2]
```diff
@@ -33,51 +33,61 @@
...
     {
-        for(int cx = start_w; cx < end_w; cx++)
+        for(uint cx = start_w; cx < end_w; cx++)
         {
             if((im_h - cy * stride_h) % dilation_h == 0 && (im_w - cx * stride_w) % dilation_w == 0)
             {
-                int col_off_y = cy + (((im_h - cy * stride_h) / dilation_h) * wei_w * col_h);
-                int col_off_x = cx + (((im_w - cx * stride_w) / dilation_w) * col_w * col_h);
+                index_t col_off_y = cy + (((im_h - cy * stride_h) / dilation_h) * wei_w * col_h);
+                index_t col_off_x = cx + (((im_w - cx * stride_w) / dilation_w) * col_w * col_h);
```
Comment (atamazov) [inline src/kernels/MIOpenCol2Im2d.cl:90]: Ditto

File: src/kernels/MIOpenCol2Im2d.cl [hunk 3]
```diff
@@ -33,51 +33,61 @@
...
+                        const uint dilation_h,
+                        const uint dilation_w,
+                        const uint height,
+                        const uint width,
+                        global _FLOAT* im,
+                        const uint im_offset)
 {
     global _FLOAT* im_off = im + im_offset;
-    int gid               = (int)get_global_id(0);
+    uint gid              = (uint)get_global_id(0);
```
Comment (atamazov) [inline src/kernels/MIOpenCol2Im2d.cl:63]: All offset math switched to unsigned domain, which doubles the upper limit for free.

File: src/kernels/MIOpenCol2Im3d.cl [hunk 4]
```diff
@@ -33,88 +33,92 @@
...
+                        const uint dilation_w,
+                        const uint depth,
+                        const uint height,
+                        const uint width,
+                        global _FLOAT* im,
+                        const unsigned long im_offset)
 {
     global _FLOAT* im_off = im + im_offset;
-    int gid               = (int)get_global_id(0);
+    uint gid              = (uint)get_global_id(0);
```
Comment (atamazov) [inline src/kernels/MIOpenCol2Im3d.cl:63]: Ditto in 3D Col2Im

File: src/ocl/utilocl.cpp [hunk 5]
```diff
@@ -468,6 +470,17 @@ float Col2Im2dGPU(const Handle& handle,
         const std::vector<size_t> vgd{global_threads, 1, 1};
         const std::vector<size_t> vld{std::min(WG_SIZE, global_threads), 1, 1};
 
+        auto Is64BitIndexRequired = [&]() -> int {
+            const auto im_ch_max     = static_cast<size_t>(global_threads) / (in_w * in_h);
+            const auto ch_offset_max = im_ch_max * out_w * out_h * wei_w * wei_h;
+            MIOPEN_LOG_T("global_threads, out_h, out_w, wei_h, wei_w = "
+                         << '{' << global_threads << ',' << out_h << ',' << out_w << ',' << wei_h
+                         << ',' << wei_w << '}' << " ch_offset_max = " << ch_offset_max);
+            return (ch_offset_max > 0xffffffffULL) ? 1 : 0;
```
Comment (atamazov) [inline src/ocl/utilocl.cpp]: Here we reproduce the worst-case computations in the kernel, thus precisely defining the need in 64 bits.

**Review by atamazov — COMMENTED** [review]
Hints for reviewers.

**Comment by bghimireamd** [comment]
what about https://github.com/ROCm/MIOpen/blob/develop/src/ocl/utilocl.cpp#L43. This `int` becomes bottle neck in middle. Also, https://github.com/ROCm/MIOpen/blob/develop/src/ocl/utilocl.cpp#L84

File: src/kernels/MIOpenCol2Im2d.cl [hunk 6]
```diff
@@ -33,51 +33,61 @@
...
+#error "MIOPEN_USE_64BIT_INDEX must be defined"
+#endif
+
+#if MIOPEN_USE_64BIT_INDEX
+typedef ulong index_t;
+#else
+typedef uint index_t;
+#endif
+
+__kernel void Col2Im2dU(global _FLOAT* col,
```
Comment (atamazov) [inline src/kernels/MIOpenCol2Im2d.cl:46]: Kernel is renamed

File: src/kernels/MIOpenCol2Im3d.cl [hunk 7]
```diff
@@ -33,88 +33,92 @@
...
-                       const int depth,
-                       const int height,
-                       const int width,
-                       global _FLOAT* im,
-                       const unsigned long im_offset)
+#ifndef MIOPEN_USE_64BIT_INDEX
+#error "MIOPEN_USE_64BIT_INDEX must be defined"
+#endif
+
+__kernel void Col2Im3dU(global _FLOAT* col,
```
Comment (atamazov) [inline src/kernels/MIOpenCol2Im3d.cl:40]: Ditto.

**Review by atamazov — COMMENTED** [review]
Hints for reviewers II

**Comment by atamazov** [comment]
@bghimireamd 
> what about https://github.com/ROCm/MIOpen/blob/develop/src/ocl/utilocl.cpp#L43. This `int` becomes bottle neck in middle. Also, https://github.com/ROCm/MIOpen/blob/develop/src/ocl/utilocl.cpp#L84

Oops, we also need to fix Im2Col, thanks for reminding. I'll look there ASAP. I prefer fixing this issue by a single PR.

/cc @junliume

**Comment by atamazov** [comment]
@bghimireamd Yes, it must be passed as size_t as this is a product of 4 (2d) or 5 (3d) dimensions. Let me look how this can be fixed. Maybe I'll just borrow your fix for this.

**Comment by atamazov** [comment]
@junliume @bghimireamd @JehandadKhan Ready for review, testing & merging.
- PR description updated.
- The potential issues related to 3D Im2Col remain. But this is not urgent AFAICS. Since the fix is ​​non-trivial, I plan to provide the fix separately.

**Comment by atamazov** [comment]
Examples of resolved issues added to the PR description.

**Comment by atamazov** [comment]
@junliume @bghimireamd @JehandadKhan https://github.com/ROCm/MIOpen/labels/TESTING_CI_PASSED

Can somebody check if #3076 is resolved now with this PR, thanks.

**Comment by bghimireamd** [comment]
LGTM. I tested GemmFwdRest, GemmBwdRest, and GemmWrwUniversal, and none of them showed memory faults with the configurations that were causing memory faults before.

` MIOPEN_DEBUG_DISABLE_FIND_DB=1  MIOPEN_DEBUG_FIND_ONLY_SOLVER=GemmWrwUniversal   ./bin/MIOpenDriver conv -n 2048 -c 64 -H 224 -W 224 -D 224 -k 64 -y 3 -x 3 -z 3 -p 1 -q 1 -u 1 -v 1
-l 1 -j 1 -m conv -g 1 -F 4 -t 1 -V 0`

` MIOPEN_DEBUG_DISABLE_FIND_DB=1  MIOPEN_DEBUG_FIND_ONLY_SOLVER=GemmBwdRest   ./bin/MIOpenDriver conv -n 2048 -c 64 -H 224 -W 224 -D 224 -k 64 -y 3 -x 3 -z 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 2 -t 1 -V 0` 

`MIOPEN_DEBUG_DISABLE_FIND_DB=1  MIOPEN_DEBUG_FIND_ONLY_SOLVER=GemmFwdRest   ./bin/MIOpenDriver conv -n 2048 -c 64 -H 224 -W 224 -D 224 -k 64 -y 3 -x 3 -z 3 -p 1 -q 1 -u 1 -v 1 -l 1
 -j 1 -m conv -g 1 -F 1 -t 1 -V 0`

**Review by bghimireamd — APPROVED** [review]

**Comment by atamazov** [comment]
@bghimireamd Thanks! However to be 100% sure, we should run the vgg16 model on 8 GPU concurrently, as specified in #3076. Unfortunately I do not have access to a eight-head monster currently ;) Maybe @junliume can provide you with the necessary instructions about this?

**Review by junliume — APPROVED** [review]

**Comment by bghimireamd** [comment]
> @bghimireamd Thanks! However to be 100% sure, we should run the vgg16 model on 8 GPU concurrently, as specified in #3076. Unfortunately I do not have access to a eight-head monster currently ;) Maybe @junliume can provide you with the necessary instructions about this?

I ran below commands  and do not see mem fault
   ./bin/MIOpenDriver conv -n 1024 -c 3 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 64 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 64 -H 112 -W 112 -k 128 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 128 -H 112 -W 112 -k 128 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 128 -H 56 -W 56 -k 256 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 256 -H 56 -W 56 -k 256 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 256 -H 28 -W 28 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 512 -H 28 -W 28 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 512 -H 14 -W 14 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 512 -H 14 -W 14 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 2 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 512 -H 14 -W 14 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 4 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 512 -H 28 -W 28 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 2 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 512 -H 28 -W 28 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 4 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 256 -H 28 -W 28 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 2 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 256 -H 28 -W 28 -k 512 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 4 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 256 -H 56 -W 56 -k 256 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 2 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 256 -H 56 -W 56 -k 256 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 4 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 128 -H 56 -W 56 -k 256 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 2 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 128 -H 56 -W 56 -k 256 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 4 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 128 -H 112 -W 112 -k 128 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 2 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 128 -H 112 -W 112 -k 128 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 4 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 64 -H 112 -W 112 -k 128 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 2 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 64 -H 112 -W 112 -k 128 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 4 -t 1 -V 0
   ./bin/MIOpenDriver conv -n 1024 -c 64 -H 224 -W 224 -k 64 -y 3 -x 3 -p 1 -q 1 -u 1 -v 1 -l 1 -j 1 -m conv -g 1 -F 2 -t 1 -V 0

**Pull request merged**